File: python/nonogram.py
import enum
import random


class Pixel(enum.Enum):
    white = 0
    black = 1


class Nonogram:
    def __init__(self, width=5, height=5, seed=None):
        self.width = width
        self.height = height
        if seed is not None:
            self.seed = seed
        else:
            self.seed = random.randint(0, 2 ** (width * height))
        self.pixels = self._generate_from_seed(self.seed)
        self.hints_vertical, self.hints_horizontal = self._calculate_hints()

    def _generate_from_seed(self, seed):
        pixels = []
        for y in range(self.height):
            pixels.append([])
            for x in range(self.width):
                if (bin(seed)[2:].zfill(self.height * self.width))[y * self.width + x] == str(1):
                    pixels[y].append(Pixel.black)
                else:
                    pixels[y].append(Pixel.white)
        return pixels

    def _calculate_hints(self):
        hints_y = []
        for x in range(self.width):
            hints_y.append([])
            for y in range(self.height):
                if self.pixels[y][x] == Pixel.black:
                    if y == 0 or self.pixels[y - 1][x] == Pixel.white:
                        hints_y[x].append(1)
                    else:
                        hints_y[x][-1] = hints_y[x][-1] + 1
            if len(hints_y[x]) == 0:
                hints_y[x].append(0)

        hints_x = []
        for y in range(self.height):
            hints_x.append([])
            for x in range(self.width):
                if self.pixels[y][x] == Pixel.black:
                    if x == 0 or self.pixels[y][x - 1] == Pixel.white:
                        hints_x[y].append(1)
                    else:
                        hints_x[y][-1] = hints_x[y][-1] + 1
            if len(hints_x[y]) == 0:
                hints_x[y].append(0)

        return hints_y, hints_x

    def print(self):
        hint_column_number = max(len(hints) for hints in self.hints_horizontal)
        hint_row_number = max(len(hints) for hints in self.hints_vertical)

        for hint_row in range(hint_row_number):
            print(' ' * hint_column_number, end='')
            for hint_column in range(self.width):
                if hint_row < len(self.hints_vertical[hint_column]):
                    print(self.hints_vertical[hint_column][hint_row], end='')
                else:
                    print(' ', end='')
            print('\n', end='')

        for y in range(self.height):

            for hint_column in range(hint_column_number):
                if hint_column < len(self.hints_horizontal[y]):
                    print(self.hints_horizontal[y][hint_column], end='')
                else:
                    print(' ', end='')

            for hint_column in range(self.width):
                if self.pixels[y][hint_column] == Pixel.black:
                    print('█', end='')
                else:
                    print(' ', end='')
            print('\n', end='')

    def _calculate_groups_content_score(self, self_line, neighbour_line):
        score = 0
        for x in range(min(len(self_line), len(neighbour_line))):
            if self_line[x] == neighbour_line[x]:
                score += 1

        return score

    def calculate_fitness_score(self, candidate):
        if candidate.height != self.height or candidate.width != self.width:
            return 0

        fitness_score = 0
        for y in range(self.height):
            # matching lengths groups count
            fitness_score += max(0, len(self.hints_vertical[y]) - abs(
                len(self.hints_vertical[y]) - len(candidate.hints_vertical[y])))
            # matching groups sums
            fitness_score += max(0, sum(self.hints_vertical[y]) - abs(
                sum(self.hints_vertical[y]) - sum(candidate.hints_vertical[y])))
            # matching groups content
            fitness_score += self._calculate_groups_content_score(self.hints_vertical[y], candidate.hints_vertical[y])

        for x in range(self.width):
            # matching lengths groups count
            fitness_score += max(0, len(self.hints_horizontal[x]) - abs(
                len(self.hints_horizontal[x]) - len(candidate.hints_horizontal[x])))
            # matching groups sums
            fitness_score += max(0, sum(self.hints_horizontal[x]) - abs(
                sum(self.hints_horizontal[x]) - sum(candidate.hints_horizontal[x])))
            # matching groups content
            fitness_score += self._calculate_groups_content_score(self.hints_horizontal[x],
                                                                  candidate.hints_horizontal[x])

        return fitness_score

    def get_neighbours(self):
        seeds = [self.seed ^ (1 << bit) for bit in range(self.width * self.height)]
        neighbours = [Nonogram(self.width, self.height, seed=seed) for seed in seeds]
        return neighbours

File: cpp/nonogram.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>
#include "nonogram.h"
#include "utils/random.h"

Nonogram::Nonogram(int Width, int Height) : width(Width), height(Height) {
  Random<unsigned long long> random;
  unsigned long long generated = random.generate(0, (unsigned long long)pow(2, width * height));
  this->seed = number_to_seed(generated);
  calculate_hints();
}
Nonogram::Nonogram(int Width, int Height, std::vector<bool> Seed)
	: width(Width), height(Height), seed(Seed) {
  calculate_hints();
}
Nonogram::Nonogram(int Width, int Height, unsigned long long Seed)
	: width(Width), height(Height), seed(number_to_seed(Seed)) {
  calculate_hints();
}

bool Nonogram::operator==(const Nonogram &Rhs) const {
  return width == Rhs.width &&
	  height == Rhs.height &&
	  hints_vertical == Rhs.hints_vertical &&
	  hints_horizontal == Rhs.hints_horizontal &&
	  seed == Rhs.seed;
}
bool Nonogram::operator!=(const Nonogram &Rhs) const {
  return !(Rhs == *this);
}
std::vector<Nonogram> Nonogram::get_neighbours() {
  std::vector<Nonogram> neighbours;
  for (int i = 0; i < this->seed.size(); i++) {
	std::vector<bool> new_seed = this->seed;
	new_seed[i] = !new_seed[i];
	neighbours.push_back(Nonogram(this->width, this->height, new_seed));
  }
  return neighbours;
}

int Nonogram::calculate_fitness_score(Nonogram candidate) {
  if (candidate.height != this->height || candidate.width != this->width) {
	return 0;
  }

  int fitness_score = 0;
  for (int y = 0; y < this->height; y++) {
	// matching lengths groups count
	fitness_score += std::max(0, (int)this->hints_vertical[y].size() - abs(
		(int)(this->hints_vertical[y].size() - candidate.hints_vertical[y].size())));
	// matching groups sums
	fitness_score +=
		std::max(0, std::accumulate(this->hints_vertical[y].begin(), this->hints_vertical[y].end(), 0) - abs(
			std::accumulate(this->hints_vertical[y].begin(), this->hints_vertical[y].end(), 0)
				- std::accumulate(candidate.hints_vertical[y].begin(), candidate.hints_vertical[y].end(), 0)));
	// matching groups content
	fitness_score += calculate_groups_content_score(this->hints_vertical[y], candidate.hints_vertical[y]);
  }

  for (int x = 0; x < this->width; x++) {
	// matching lengths groups count
	fitness_score += std::max(0, (int)this->hints_horizontal[x].size() - abs(
		(int)(this->hints_horizontal[x].size() - candidate.hints_horizontal[x].size())));
	// matching groups sums
	fitness_score +=
		std::max(0, std::accumulate(this->hints_horizontal[x].begin(), this->hints_horizontal[x].end(), 0) - abs(
			std::accumulate(this->hints_horizontal[x].begin(), this->hints_horizontal[x].end(), 0)
				- std::accumulate(candidate.hints_horizontal[x].begin(), candidate.hints_horizontal[x].end(), 0)));
	// matching groups content
	fitness_score += calculate_groups_content_score(this->hints_horizontal[x], candidate.hints_horizontal[x]);
  }

  return fitness_score;
}

int Nonogram::get_max_score() {
  if (this->max_score > 0) {
	return this->max_score;
  } else {
	this->max_score = calculate_fitness_score(*this);
	return this->max_score;
  }
}

void Nonogram::print() {
  int hint_column_number = 0;
  int hint_row_number = 0;
  for (int i = 0; i < this->hints_horizontal.size(); i++) {
	if (this->hints_horizontal[i].size() > hint_column_number)
	  hint_column_number = this->hints_horizontal[i].size();
  }
  for (int i = 0; i < this->hints_vertical.size(); i++) {
	if (this->hints_vertical[i].size() > hint_row_number)
	  hint_row_number = this->hints_vertical[i].size();
  }

  for (int hint_row = 0; hint_row < hint_row_number; hint_row++) {
	for (int i = 0; i < hint_column_number; i++) {
	  std::cout << ' ';
	}
	for (int hint_column = 0; hint_column < this->width; hint_column++) {
	  if (hint_row < this->hints_vertical[hint_column].size()) {
		std::cout << this->hints_vertical[hint_column][hint_row];
	  } else {
		std::cout << ' ';
	  }
	}
	std::cout << std::endl;
  }

  for (int y = 0; y < this->height; y++) {

	for (int hint_column = 0; hint_column < hint_column_number; hint_column++) {
	  if (hint_column < this->hints_horizontal[y].size()) {
		std::cout << this->hints_horizontal[y][hint_column];
	  } else {
		std::cout << ' ';
	  }
	}

	for (int x = 0; x < this->width; x++) {
	  if (this->seed[y * this->width + x]) {
		std::cout << "█";
	  } else {
		std::cout << ' ';
	  }
	}
	std::cout << std::endl;
  }
}

std::vector<bool> Nonogram::number_to_seed(unsigned long long hex) {
  std::vector<bool> new_seed;
  for (int i = 0; i < this->width * this->height; i++) {
	new_seed.push_back(hex & 1);
	hex >>= 1;
  }
  return new_seed;
}

void Nonogram::calculate_hints() {
  std::vector<std::vector<int>> hints_y;
  for (int x = 0; x < this->width; x++) {
	std::vector<int> vector;
	hints_y.push_back(vector);
	for (int y = 0; y < this->height; y++) {
	  if (this->seed[y * this->width + x] == true) {
		if (y == 0 || this->seed[(y - 1) * this->width + x] == false) {
		  hints_y[x].push_back(1);
		} else {
		  hints_y[x][hints_y.size() - 1] = hints_y[x][hints_y.size() - 1] + 1;
		}
	  }
	}
	if (hints_y[x].empty()) {
	  hints_y[x].push_back(0);
	}
  }

  std::vector<std::vector<int>> hints_x;
  for (int y = 0; y < this->height; y++) {
	std::vector<int> vector;
	hints_x.push_back(vector);
	for (int x = 0; x < this->width; x++) {
	  if (seed[y * this->width + x] == true) {
		if (x == 0 || seed[y * this->width + x - 1] == false) {
		  hints_x[y].push_back(1);
		} else {
		  hints_x[y][hints_x.size() - 1] = hints_x[y][hints_x.size() - 1] + 1;
		}
	  }
	}
	if (hints_x[y].empty()) {
	  hints_x[y].push_back(0);
	}
  }

  this->hints_vertical = hints_y;
  this->hints_horizontal = hints_x;
}

int Nonogram::calculate_groups_content_score(std::vector<int> self_line, std::vector<int> candidate_line) {
  int score = 0;
  for (int x = 0; x < std::min(self_line.size(), candidate_line.size()); x++) {
	if (self_line[x] == candidate_line[x]) {
	  score += 1;
	}
  }
  return score;
}

